text/0001-catalogs.md

```diff
@@ -0,0 +1,346 @@
+
+As a result, multiple versions of the same dependency appear over time. This can cause different flavors of surprising problems.
+
+- In projects that bundle dependencies, multiple versions inflate the size of the final result deployed to users.
+- Differing versions result in multiple copies that may not interact well at runtime, especially if features like [`Symbol()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) are used. For example, [React hooks will error if a component is rendered with a different copy of React in the same app](https://reactjs.org/warnings/invalid-hook-call-warning.html#duplicate-react).
+- For TypeScript, multiple copies of the same `@types` package causes compile errors from mismatching definitions. The compiler diagnostic for this is usually: *"argument of type `Foo` is not assignable to parameter of type `Foo`"*. For developers that have seen this before, they may realize this diagnostic is due to a dependency version mismatch. For developers new to TypeScript, *"`Foo` is not assignable to `Foo`"* is very confusing.
+
+While there are situations differing versions are intentional, this is more often accidental. Multiple differing versions arise from not reviewing `pnpm-lock.yaml` file changes or not searching for existing dependency specifiers before adding a new one. The later is typically unwritten convention in most monorepos.
+
+### Merge Conflict Resistance
+
+In addition to reducing the likelihood of multiple versions of the same dependency in a monorepo, the new `catalog:` protocol reduces merge conflicts. Any `package.json` files using the catalog protocol to declare a dependency do not need to be edited when changing the version of that dependency. This side steps `package.json` merge conflicts by avoiding edits to them in the first place.
+
+**Merge conflict resistance is the a primary motivator for introducing catalogs as a first-class feature to pnpm.** This is only possible through a new specifier protocol. See [_What kinds of merge conflicts are avoided?_](#what-kinds-of-merge-conflicts-are-avoided) for details.
+
+## Detailed Explanation
+
+### Configuring Catalogs
+
+Catalogs are configured on the root `package.json` of a workspace and available to all workspace packages.
+
+The keys of the `pnpm.catalogs` map will be available for reference through `catalog:<name>`.
+
+```json5
+// package.json
+{
+  "name": "monorepo-root",
+  "pnpm": {
+    "catalogs": {
+
+      // Can be referenced through "catalog:default" or "catalog:"
+      "default": {
+        "jest": "^29.6.1",
+        "redux": "^4.2.0",
+        "react-redux": "^8.0.0"
+      },
+
+      // Can be referenced through "catalog:react17"
+      "react17": {
+        "react": "^17.0.2",
+        "react-dom": "^17.0.2"
+      },
+
+      // Can be referenced through "catalog:react18"
+      "react18": {
+        "react": "^18.2.0",
+        "react-dom": "^18.2.0"
+      },
+
```

Comment:
I imagine that most people will only need `default`, so for them, having to write a `"default"` object after `"catalogs"` will feel like a hassle.

Reply:
That's a good point. I do agree. Here's another potential config structure:

```json5
// package.json
{
  // Can be referenced through `catalog:default` or `catalog:`.
  "catalog": {
    "jest": "^29.6.1",
    "redux": "^4.2.0",
    "react-redux": "^8.0.0"
  },

  "namedCatalogs": {
    // Can be referenced through "catalog:react17"
    "react17": {
      "react": "^17.0.2",
      "react-dom": "^17.0.2"
    },

    // Can be referenced through "catalog:react18"
    "react18": {
      "react": "^18.2.0",
      "react-dom": "^18.2.0"
    }
  }
}
```

We could also name these `defaultCatalog` and `catalogs`, but under the idea that most users will just have a default catalog, I'd want to keep that name shorter and just `catalog`.

Reply:
> We could also name these `defaultCatalog` and `catalogs`, but under the idea that most users will just have a default catalog, I'd lean towards keeping that name shorter and just `catalog`.

I'm fine with either way.

Reply:
Updated the document for the `catalog` and `namedCatalogs` split here: b8b4527839c086edaa617a58895525f578eee20f

Reply:
You could also put everything to catalog like this

```jsonc
// package.json
{
  // Can be referenced through `catalog:default` or `catalog:`.
  "catalog": {
    "jest": "^29.6.1",
    "redux": "^4.2.0",
    "react-redux": "^8.0.0"

    // Can be referenced through "catalog:react17"
    "react17": {
      "react": "^17.0.2",
      "react-dom": "^17.0.2"
    },

    // Can be referenced through "catalog:react18"
    "react18": {
      "react": "^18.2.0",
      "react-dom": "^18.2.0"
    }
  }
}
```

Inspired by npm's [overrides](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) syntax. If you want to use the same name for the catalog group as a dependency name, you can use `.`. For instance:

```json
{
  "catalog": {
    "jest": "^29.6.1",
    "redux": "^4.2.0",
    "react-redux": "^8.0.0"
    "react": {
      ".": "^17.0.2",
      "react-dom": "^17.0.2"
    }
}
```

Reply:
That looks cool too. I hadn't thought of that @zkochan. I like how the overrides-style syntax looks. Thinking through how it works, my only question would be whether or not users will think the overrides-style syntax implies a hierarchy when catalogs are otherwise flat?

From looking at the [`overrides` docs](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides), it looks like there is an intentional hierarchy in that case. The override for `foo` below only applies to situations when it's a dependency of `bar`.

```json
{
  "overrides": {
    "bar": {
      "foo": "1.0.0"
    }
  }
}
```

I personally don't have a strongly preference between `namedCatalogs` and the overrides syntax. @zkochan If you have a preference, I'm happy to go in that direction. Appreciate the suggestion regardless. 🙂

Reply:
I like the overrides-style more.

Reply:
I'm updating the RFC for the overrides-style and realizing there's a bit of ambiguity in this case.

> Inspired by npm's [overrides](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) syntax. If you want to use the same name for the catalog group as a dependency name, you can use `.`. For instance:
> 
> ```json
> {
>   "catalog": {
>     "jest": "^29.6.1",
>     "redux": "^4.2.0",
>     "react-redux": "^8.0.0"
>     "react": {
>       ".": "^17.0.2",
>       "react-dom": "^17.0.2"
>     }
> }
> ```

Does setting `catalog.react["."]` configure it for the named catalog, the default catalog, or both? I had originally imagined it was just for the named catalog. So the `^17.0.02` specifier above would be available for reference through:

```json
{
  "dependencies": {
    "react": "catalog:react"
  }
}
```

What happens if I configure the above for this? Does this resolve to `^17.0.02` or error?

```json
{
  "dependencies": {
    "react": "catalog:"
  }
}
```

Reply:
For now I've updated the RFC for the new nested config syntax: d15af41af2c3748e040c1d0daf79afd82dbeaf83

I'll describe the `.` case after getting clarification on whether we want that to apply to the default catalog or the named catalog.

Reply:
> Does setting catalog.react["."] configure it for the named catalog

It configures it for the default catalog.

Reply:
Added in https://github.com/pnpm/rfcs/pull/1/commits/3f7bf75f1219a160d44a48a0497b139337752508.

> It configures it for the default catalog.

I like how clean the overrides-style syntax looks, but I do worry that the specifics of how `.` configures the default catalog will be a bit confusing and too magical. My intuition from looking at the `.` syntax was that it added it to the named catalog.

Reply:
The concepts in catalogs and overrides are not equivalent at all. The group names in catalogs are not package names. I **strongly** oppose imitating overrides.

Reply:
We can remove the `.` bit and still use nesting inside `catalogs` instead of dividing to 2 fields.

Reply:
anyway, I don't care. I was fine with the original way it was with the "default" key in "catalog" and two fields is fine too.

Reply:
Instead of a `namedCatalogs`, what do you guys think about `catalog:*` as name? For example:

```json5
{
  // equivalent to "catalog:default"
  // can be referenced through "catalog:" or "catalog:default"
  "catalog": {    
    "jest": "^29.6.1",
    "redux": "^4.2.0",
    "react-redux": "^8.0.0",
  },

  // can be referenced through "catalog:react17"
  "catalog:react17": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
  },
}
```

If both `catalog` and `catalog:default` exist, an error is thrown.

Reply:
No, I don't think this one is good.

Reply:
@zkochan Is there a strong preference away from configuring this in a separate file? We previously discussed this in https://github.com/pnpm/rfcs/pull/1#discussion_r1258936332, but I'm not sure if that was a strong preference or something we do want to revisit after the initial implementation.

If it is something we'd want to revisit anyway, it may be worth another look now since there isn't an inline config we're all completely happy with.

I do think the separate file is the best option for a few reasons:

- For some monorepos, the default catalog list is going to be huge. When there's 1000+ dependencies, I think it'll be easier to manage in a separate file instead of all shoved into the root `package.json`.
- It has some precedence in dependency management tools from other languages. An example would be [Gradle's `libs.versions.toml` file](https://docs.gradle.org/current/userguide/platforms.html#sub::toml-dependencies-format)
- It makes it easier to add on new configs related to the catalog. For example, I could see a `groups` or variables-like feature being useful in the future.
  ```yaml
  groups:
    # Packages from the @typescript-eslint repo are typically versioned together.
    typescript-eslint: ^6.5.0

  versions:
    '@typescript-eslint/eslint-plugin': $typescript-eslint
    '@typescript-eslint/parser': $typescript-eslint
    react: ^18.2.0
    react-dom: ^18.2.0
  ```
- I like the idea of a `pnpm-catalog.yaml` next to the existing `pnpm-workspace.yaml` file.

I don't think the separate file would add too much implementation complexity and I suspect this is the option that most users will prefer too?

Reply:
> Instead of a namedCatalogs, what do you guys think about catalog:* as name?

@KSXGitHub I do like that this looks as clean as the overrides-style option. My only nitpick would be that it'd be hard to represent in a JSON Schema config since the keys are dynamic. I'm also unsure if the dynamic keys add config parsing complexity. It's certainly doable, but there might be some longer maintenance cost of having to read unknown keys to check whether their starting value is `catalog:`?

I `@`-mentioned Zoltan above since he and I discussed a separate file config previously, but would be happy to hear your thoughts on the external file config as well.

Reply:
@gluxon Separate file is better, I agree.

Reply:
I don't know how I feel about a separate file. I am not sure it is a great idea because we already have so many different ways to configure pnpm. But I will not reject the RFC if you decide to go with a separate file.

We might want to consider putting the new config file to a `.pnpm` directory.

Reply:
Catalogs looks like a workspace feature. I think it should be in [pnpm-workspace.yaml](https://pnpm.io/pnpm-workspace_yaml). This config file currently only has one top-level field called `packages`.

Reply:
From a solely end user experience, it feels like there's no disagreement that a separate file (or reusing `pnpm-workspace.yaml`) results in a better configuration story? Since the contention is codebase complexity and reducing the multiple ways to configure pnpm, I definitely trust @zkochan to speak to that best.

I will wait for @zkochan's thoughts on reusing the `pnpm-workspace.yaml` file since I see a few others have reacted positively to that last suggestion. I personally want to provide the best user experience but maintaining a simpler codebase is also very important.

Reply:
We will definitely need to revise the different ways we configure pnpm. But `pnpm-workspace.yaml` already exists and as @KSXGitHub the catalogs in the current state only makes sense in a workspace, so I like this idea.

Reply:
Perfect! Here's the commit modifying the RFC to specify `pnpm-workspace.yaml`. https://github.com/pnpm/rfcs/pull/1/commits/2610503e966075bb24138d6fcd1a2c51b8d452e0

I did switch back to `catalog`/`namedCatalogs` since it allows users to define a named catalog and a default catalog entry of the same name without the `.` syntax.